File: packages/dsp/crates/dsp-core/src/stream/meters.rs
//! Level metering for the mixer strips, channel meters, and headphone bus.
//!
//! Measured at the emit position rather than the render horizon, so a meter
//! never leads the audio the listener is hearing.

/// RMS and peak over one emitted block.
#[derive(Clone, Copy, Debug, Default)]
pub struct Level {
    pub rms: f64,
    pub peak: f64,
}

impl Level {
    pub fn measure(samples: &[f64]) -> Self {
        if samples.is_empty() {
            return Self::default();
        }
        let mut sum_sq = 0.0;
        let mut peak = 0.0_f64;
        for v in samples {
            sum_sq += v * v;
            peak = peak.max(v.abs());
        }
        Self {
            rms: (sum_sq / samples.len() as f64).sqrt(),
            peak,
        }
    }

    pub fn measure_f32(samples: &[f32], gain: f64) -> Self {
        if samples.is_empty() {
            return Self::default();
        }
        let mut sum_sq = 0.0;
        let mut peak = 0.0_f64;
        for v in samples {
            let scaled = *v as f64 * gain;
            sum_sq += scaled * scaled;
            peak = peak.max(scaled.abs());
        }
        Self {
            rms: (sum_sq / samples.len() as f64).sqrt(),
            peak,
        }
    }
}

/// The mastering readouts: loudness of the delivered programme and how hard
/// each dynamics stage is working, both over the same trailing window the
/// level meters use.
#[derive(Clone, Copy, Debug, Default)]
pub struct MasterMeters {
    /// EBU Tech 3341 momentary (400 ms) loudness, LKFS.
    pub momentary_lkfs: f64,
    /// EBU Tech 3341 short-term (3 s) loudness, LKFS.
    pub short_term_lkfs: f64,
    /// Deepest bus-compressor gain reduction, dB, positive downward.
    pub comp_gr_db: f64,
    /// Deepest limiter gain reduction on the mains curve, dB.
    pub limiter_gr_db: f64,
    /// Deepest limiter gain reduction on the LFE's own curve, dB.
    pub limiter_lfe_gr_db: f64,
}

/// Floats [`MasterMeters`] writes.
const MASTER_METER_FLOATS: usize = 5;

/// Everything the UI meters, in one block the host reads after each render.
#[derive(Clone, Debug, Default)]
pub struct Meters {
    /// Per stem, post-gain and pre-routing — what a mixer strip shows.
    /// Left/right pair always: a mono source has identical channels (see
    /// `StemSource`'s decode), so the host slices to one bar itself from the
    /// project's own channel count rather than the core tracking mono/stereo.
    pub stems: Vec<[Level; 2]>,
    /// One linked stem-dynamics reduction value per stem, positive downward.
    pub stem_dynamics_gr_db: Vec<f64>,
    /// One linked tame/dynamic-EQ reduction value per stem.
    pub stem_dynamic_eq_gr_db: Vec<f64>,
    /// Per speaker of the mastered bed.
    pub channels: Vec<Level>,
    /// The collapsed pair, or the first two bed channels for native output.
    pub output: [Level; 2],
    /// Loudness and gain-reduction readouts for the master strip.
    pub master: MasterMeters,
}

impl Meters {
    /// Flatten to `[rms, peak]` pairs: stems (left, right), stem reductions, then channels,
    /// then output, then [`MasterMeters`]'s own flat floats.
    pub fn write(&self, out: &mut [f32]) -> usize {
        let mut i = 0;
        let push = |level: &Level, out: &mut [f32], i: &mut usize| {
            if *i + 1 < out.len() {
                out[*i] = level.rms as f32;
                out[*i + 1] = level.peak as f32;
            }
            *i += 2;
        };
        for pair in &self.stems {
            push(&pair[0], out, &mut i);
            push(&pair[1], out, &mut i);
        }
        for value in &self.stem_dynamics_gr_db {
            if i < out.len() {
                out[i] = *value as f32;
            }
            i += 1;
        }
        for level in &self.channels {
            push(level, out, &mut i);
        }
        for level in &self.output {
            push(level, out, &mut i);
        }
        let m = &self.master;
        for value in [
            m.momentary_lkfs,
            m.short_term_lkfs,
            m.comp_gr_db,
            m.limiter_gr_db,
            m.limiter_lfe_gr_db,
        ] {
            if i < out.len() {
                out[i] = value as f32;
            }
            i += 1;
        }
        for value in &self.stem_dynamic_eq_gr_db {
            if i < out.len() { out[i] = *value as f32; }
            i += 1;
        }
        i
    }

    /// Number of floats [`write`] needs.
    pub fn len(&self) -> usize {
        2 * (self.stems.len() * 2 + self.channels.len() + 2)
            + self.stem_dynamics_gr_db.len()
            + self.stem_dynamic_eq_gr_db.len()
            + MASTER_METER_FLOATS
    }

    pub fn is_empty(&self) -> bool {
        self.len() == 0
    }
}

File: apps/web/src/features/projects/wasmEngine/meters.ts
export type MeterLevel = { rms: number; peak: number; clipped: boolean };

export const SILENT_METER_LEVEL: MeterLevel = { rms: 0, peak: 0, clipped: false };

// A sample clearing unity by a hairline still counts as clipped.
const CLIP_TOLERANCE = 1.0;

function level(rms: number, peak: number): MeterLevel {
  return { rms, peak, clipped: peak > CLIP_TOLERANCE };
}

export type MeterFrame = { position: number; meters: number[]; spectrum: number[] };

export type StemSpectrum = { level: number; centroid: number };

/** The master strip's readouts — see `MasterMeters` in the core. LKFS over
 * the delivered programme; the core's gain-reduction floats follow these two
 * on the wire and are not read. */
export type MasterMeters = {
  momentaryLkfs: number;
  shortTermLkfs: number;
};

export const SILENT_MASTER_METERS: MasterMeters = {
  momentaryLkfs: -70,
  shortTermLkfs: -70,
};

export type DecodedMeters = {
  stemLevels: Map<string, MeterLevel[]>;
  stemDynamics: Map<string, number>;
  stemDynamicEq: Map<string, number>;
  stemSpectrum: Map<string, StemSpectrum>;
  channelLevels: Map<string, MeterLevel>;
  headphoneLevels: { left: MeterLevel; right: MeterLevel };
  master: MasterMeters;
};

/**
 * Unpack one render-callback frame. The meter array is laid out as
 * `[stems…][channels…][headphone L/R][master]`, four values per stem (two per
 * channel of the stem), two per output channel, and five for the master.
 */
export function decodeMeterFrame(
  frame: MeterFrame,
  stemOrder: string[],
  stemChannelCounts: number[],
  channels: string[],
): DecodedMeters {
  const meters = frame.meters;
  const stemCount = stemOrder.length;
  const stemLevels = new Map<string, MeterLevel[]>();
  const stemDynamics = new Map<string, number>();
  const stemDynamicEq = new Map<string, number>();
  const stemSpectrum = new Map<string, StemSpectrum>();
  for (let i = 0; i < stemCount; i += 1) {
    const o = i * 4;
    const bars = [level(meters[o] ?? 0, meters[o + 1] ?? 0)];
    if ((stemChannelCounts[i] ?? 1) >= 2) bars.push(level(meters[o + 2] ?? 0, meters[o + 3] ?? 0));
    stemLevels.set(stemOrder[i], bars);
    const s = i * 2;
    stemSpectrum.set(stemOrder[i], {
      level: frame.spectrum[s] ?? 0,
      centroid: frame.spectrum[s + 1] ?? 0,
    });
  }

  const dynamicsBase = stemCount * 4;
  for (let i = 0; i < stemCount; i += 1) stemDynamics.set(stemOrder[i], meters[dynamicsBase + i] ?? 0);
  const channelLevels = new Map<string, MeterLevel>();
  const base = dynamicsBase + stemCount;
  for (let i = 0; i < channels.length; i += 1) {
    channelLevels.set(channels[i], level(meters[base + i * 2] ?? 0, meters[base + i * 2 + 1] ?? 0));
  }

  const outBase = base + channels.length * 2;
  const masterBase = outBase + 4;
  const dynamicEqBase = masterBase + 5;
  for (let i = 0; i < stemCount; i += 1) stemDynamicEq.set(stemOrder[i], meters[dynamicEqBase + i] ?? 0);
  return {
    stemLevels,
    stemDynamics,
    stemDynamicEq,
    stemSpectrum,
    channelLevels,
    headphoneLevels: {
      left: level(meters[outBase] ?? 0, meters[outBase + 1] ?? 0),
      right: level(meters[outBase + 2] ?? 0, meters[outBase + 3] ?? 0),
    },
    master: {
      momentaryLkfs: meters[masterBase] ?? -70,
      shortTermLkfs: meters[masterBase + 1] ?? -70,
    },
  };
}
